spring-kafka/src/main/java/org/springframework/kafka/support/converter/BatchMessagingMessageConverter.java

```diff
@@ -0,0 +1,147 @@
+ * @author Dariusz Szablinski
+ * @since 1.1
+ */
+public class BatchMessagingMessageConverter implements BatchMessageConverter {
+
+	private boolean generateMessageId = false;
+
+	private boolean generateTimestamp = false;
+
+	/**
+	 * Generate {@link Message} {@code ids} for produced messages. If set to {@code false},
+	 * will try to use a default value. By default set to {@code false}.
+	 * @param generateMessageId true if a message id should be generated
+	 */
+	public void setGenerateMessageId(boolean generateMessageId) {
+		this.generateMessageId = generateMessageId;
+	}
+
+	/**
+	 * Generate {@code timestamp} for produced messages. If set to {@code false}, -1 is
+	 * used instead. By default set to {@code false}.
+	 * @param generateTimestamp true if a timestamp should be generated
+	 */
+	public void setGenerateTimestamp(boolean generateTimestamp) {
+		this.generateTimestamp = generateTimestamp;
+	}
+
+	@Override
+	public Message<?> toMessage(List<ConsumerRecord<?, ?>> records, Acknowledgment acknowledgment,
+			Type type) {
+		KafkaMessageHeaders kafkaMessageHeaders = new KafkaMessageHeaders(this.generateMessageId,
+				this.generateTimestamp);
+
+		Map<String, Object> rawHeaders = kafkaMessageHeaders.getRawHeaders();
+		List<Object> payloads = new ArrayList<>();
+		List<Object> keys = new ArrayList<>();
+		List<String> topics = new ArrayList<>();
+		List<Integer> partitions = new ArrayList<>();
+		List<Long> offsets = new ArrayList<>();
+		rawHeaders.put(KafkaHeaders.RECEIVED_MESSAGE_KEY, keys);
+		rawHeaders.put(KafkaHeaders.RECEIVED_TOPIC, topics);
+		rawHeaders.put(KafkaHeaders.RECEIVED_PARTITION_ID, partitions);
+		rawHeaders.put(KafkaHeaders.OFFSET, offsets);
+
+		if (acknowledgment != null) {
+			rawHeaders.put(KafkaHeaders.ACKNOWLEDGMENT, acknowledgment);
+		}
+
+		for (ConsumerRecord<?, ?> record : records) {
+			payloads.add(extractAndConvertValue(record, type));
+			keys.add(record.key());
+			topics.add(record.topic());
+			partitions.add(record.partition());
+			offsets.add(record.offset());
+		}
+		return MessageBuilder.createMessage(payloads, kafkaMessageHeaders);
+	}
+
+	@Override
+	public List<ProducerRecord<?, ?>> fromMessage(Message<?> message, String defaultTopic) {
+		throw new UnsupportedOperationException();
+	}
+
+	/**
+	 * Subclasses can convert the payload; by default, it's sent unchanged to Kafka.
+	 * @param message the message.
+	 * @return the payload.
+	 */
+	protected Object convertPayload(Message<?> message) {
+		Object payload = message.getPayload();
+		if (payload instanceof KafkaNull) {
+			return null;
+		}
+		else {
+			return payload;
+		}
+	}
+
+	/**
+	 * Subclasses can convert the value; by default, it's returned as provided by Kafka.
+	 * @param record the record.
+	 * @param type the required type.
+	 * @return the value.
+	 */
+	protected Object extractAndConvertValue(ConsumerRecord<?, ?> record, Type type) {
+		return record.value() == null ? KafkaNull.INSTANCE : record.value();
```

Comment:
The `type` is out of use, but I guess it is just a copy/paste from the `MessagingMessageConverter`.
I wonder if we could use `ConversionService` to return the value for the desired `type`...

OTOH since we are here in the `batching` can't we bypass `KafkaNull.INSTANCE` and just return `null` as is for the `payloads` item?